lib/ToolbarPanelHub.jsm

```diff
@@ -132,19 +139,31 @@ class _ToolbarPanelHub {
 
     if (messages && !container.querySelector(".whatsNew-message")) {
       let previousDate = 0;
-      for (let { content } of messages) {
+      for (let message of messages) {
         container.appendChild(
-          this._createMessageElements(win, doc, content, previousDate)
+          this._createMessageElements(win, doc, message, previousDate)
         );
-        previousDate = content.published_date;
+        previousDate = message.content.published_date;
       }
     }
 
-    // TODO: TELEMETRY
     this._onPanelHidden(win);
+
+    // Panel impressions are not associated with one particular message
+    // but with a set of messages. We concatenate message ids and send them
+    // back for every impression.
+    const eventId = { id: messages.map(({ id }) => id).join(",") };
```

Comment:
Could you sort all the message IDs before joining them together? That'll make the analysis (such `SELECT distinct(message_id)...`) easier.